BUG: sql different offsets (1/2 modify infer_dtype)

ThibTrip opened the pull request:
- [x] closes #30207
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

# Explanations

For to_sql no errors are raised anymore when we have Series containing datetimes with timezone and different offsets (unlike in GH30207) and the data is saved as expected (I tested with postgres which does not save offsets anyways).

Concerning read_sql however there are several points to discuss:

1. if we want to have datetime64 instead of "object" (as described in [GH30207](https://github.com/pandas-dev/pandas/issues/30207)) we lose the offsets. So this would be a breaking change and would require first a warning in the next version then we implement it in the version after, right?

2. is this patch I made acceptable?

The probem is that read_sql uses pd.DataFrame.from_records and it reads datetimes with different offsets as "object" instead of datetime64[ns,...] (e.g. datetime64[ns, psycopg2.tz.FixedOffsetTimezone(offset=60, name=None)]).

If we wanted to modify this behavior we would have to modify the C extension _libs.lib. I don't know how to help with that and I am not sure we want to do this. Even if we could pack the datetimes with differents offsets into a Series with a datetime64 dtype the methods of the .dt accessor are not compatible anyways: currently a ValueError is thrown if one attempts to use the accessor in this case (see traceback in the example in "Before PR" below).



# Before PR


```python
import datetime
import psycopg2
import sqlalchemy
import pandas as pd
from pandas.io.sql import read_sql, to_sql

# CREATE ENGINE
# engine = sqlalchemy.create_engine("CONNECTION_STRING")

# create datetimes with timezone that have different offsets
data = [[datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60, name=None))],
        [datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=120, name=None))]]

df = pd.DataFrame(data, columns = ['ts_tz'])

# save data
to_sql(frame = df, 
       name = 'timezone_diff_offsets_test',
       con = engine,
       schema = 'public',
       if_exists = 'replace',
       index = False)
```

<details>
 <summary>Traceback</summary>


```python-traceback

    ---------------------------------------------------------------------------

    ValueError                                Traceback (most recent call last)

    ~/GitHub/pandas/pandas/core/arrays/datetimes.py in objects_to_datetime64ns(data, dayfirst, yearfirst, utc, errors, require_iso8601, allow_object)
       1968         try:
    -> 1969             values, tz_parsed = conversion.datetime_to_datetime64(data)
       1970             # If tzaware, these values represent unix timestamps, so we


    ~/GitHub/pandas/pandas/_libs/tslibs/conversion.pyx in pandas._libs.tslibs.conversion.datetime_to_datetime64()


    ValueError: Array must be all same time zone

    
    During handling of the above exception, another exception occurred:


    ValueError                                Traceback (most recent call last)

    <ipython-input-2-a66775d04fe3> in <module>
         20        schema = 'public',
         21        if_exists = 'replace',
    ---> 22        index = False)
    

    ~/GitHub/pandas/pandas/io/sql.py in to_sql(frame, name, con, schema, if_exists, index, index_label, chunksize, dtype, method)
        519         chunksize=chunksize,
        520         dtype=dtype,
    --> 521         method=method,
        522     )
        523 


    ~/GitHub/pandas/pandas/io/sql.py in to_sql(self, frame, name, if_exists, index, index_label, schema, chunksize, dtype, method)
       1318             index_label=index_label,
       1319             schema=schema,
    -> 1320             dtype=dtype,
       1321         )
       1322         table.create()


    ~/GitHub/pandas/pandas/io/sql.py in __init__(self, name, pandas_sql_engine, frame, index, if_exists, prefix, index_label, schema, keys, dtype)
        620         if frame is not None:
        621             # We want to initialize based on a dataframe
    --> 622             self.table = self._create_table_setup()
        623         else:
        624             # no data provided, read-only mode


    ~/GitHub/pandas/pandas/io/sql.py in _create_table_setup(self)
        869         from sqlalchemy import Table, Column, PrimaryKeyConstraint
        870 
    --> 871         column_names_and_types = self._get_column_names_and_types(self._sqlalchemy_type)
        872 
        873         columns = [


    ~/GitHub/pandas/pandas/io/sql.py in _get_column_names_and_types(self, dtype_mapper)
        861         column_names_and_types += [
        862             (str(self.frame.columns[i]), dtype_mapper(self.frame.iloc[:, i]), False)
    --> 863             for i in range(len(self.frame.columns))
        864         ]
        865 


    ~/GitHub/pandas/pandas/io/sql.py in <listcomp>(.0)
        861         column_names_and_types += [
        862             (str(self.frame.columns[i]), dtype_mapper(self.frame.iloc[:, i]), False)
    --> 863             for i in range(len(self.frame.columns))
        864         ]
        865 


    ~/GitHub/pandas/pandas/io/sql.py in _sqlalchemy_type(self, col)
        971             # timezone information
        972             try:
    --> 973                 if col.dt.tz is not None:
        974                     return TIMESTAMP(timezone=True)
        975             except AttributeError:


    ~/GitHub/pandas/pandas/core/accessor.py in _getter(self)
         83         def _create_delegator_property(name):
         84             def _getter(self):
    ---> 85                 return self._delegate_property_get(name)
         86 
         87             def _setter(self, new_values):


    ~/GitHub/pandas/pandas/core/indexes/accessors.py in _delegate_property_get(self, name)
         60         from pandas import Series
         61 
    ---> 62         values = self._get_values()
         63 
         64         result = getattr(values, name)


    ~/GitHub/pandas/pandas/core/indexes/accessors.py in _get_values(self)
         51                 return PeriodArray(data, copy=False)
         52             if is_datetime_arraylike(data):
    ---> 53                 return DatetimeIndex(data, copy=False, name=self.name)
         54 
         55         raise TypeError(


    ~/GitHub/pandas/pandas/core/indexes/datetimes.py in __new__(cls, data, freq, tz, normalize, closed, ambiguous, dayfirst, yearfirst, dtype, copy, name)
        266             dayfirst=dayfirst,
        267             yearfirst=yearfirst,
    --> 268             ambiguous=ambiguous,
        269         )
        270 


    ~/GitHub/pandas/pandas/core/arrays/datetimes.py in _from_sequence(cls, data, dtype, copy, tz, freq, dayfirst, yearfirst, ambiguous)
        418             dayfirst=dayfirst,
        419             yearfirst=yearfirst,
    --> 420             ambiguous=ambiguous,
        421         )
        422 


    ~/GitHub/pandas/pandas/core/arrays/datetimes.py in sequence_to_dt64ns(data, dtype, copy, tz, dayfirst, yearfirst, ambiguous)
       1864             #  or M8[ns] to denote wall times
       1865             data, inferred_tz = objects_to_datetime64ns(
    -> 1866                 data, dayfirst=dayfirst, yearfirst=yearfirst
       1867             )
       1868             tz = maybe_infer_tz(tz, inferred_tz)


    ~/GitHub/pandas/pandas/core/arrays/datetimes.py in objects_to_datetime64ns(data, dayfirst, yearfirst, utc, errors, require_iso8601, allow_object)
       1972             return values.view("i8"), tz_parsed
       1973         except (ValueError, TypeError):
    -> 1974             raise e
       1975 
       1976     if tz_parsed is not None:


    ~/GitHub/pandas/pandas/core/arrays/datetimes.py in objects_to_datetime64ns(data, dayfirst, yearfirst, utc, errors, require_iso8601, allow_object)
       1963             dayfirst=dayfirst,
       1964             yearfirst=yearfirst,
    -> 1965             require_iso8601=require_iso8601,
       1966         )
       1967     except ValueError as e:


    ~/GitHub/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.array_to_datetime()


    ~/GitHub/pandas/pandas/_libs/tslib.pyx in pandas._libs.tslib.array_to_datetime()


    ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True


```

</details>


```python
# save datetimes with UTC timezone in different periods of the year
## create test data
data = [[datetime.datetime(2019, 11, 14, 15, 12, 0, tzinfo = datetime.timezone.utc)],
        [datetime.datetime(2019, 8, 7, 13, 37, 4,  tzinfo = datetime.timezone.utc)]]

df = pd.DataFrame(data, columns = ['ts_tz'])

## save test data
to_sql(frame = df, 
       name = 'timezone_diff_offsets_test',
       con = engine,
       schema = 'public',
       if_exists = 'replace',
       index = False)

# upon reading if you have for instance "Europe/Paris"
# configured as timezone on your local machine then you get datetimes with timezone
# that have different offsets (because they are in different periods of the year)
# due to this you have "object" dtype and not datetime64[...]
df = read_sql('SELECT * FROM timezone_diff_offsets_test', engine)
print(df,'\n')
print(df.dtypes)
```

                           ts_tz
    0  2019-11-14 16:12:00+01:00
    1  2019-08-07 15:37:04+02:00 
    
    ts_tz    object
    dtype: object


# After PR


```python
import datetime
import psycopg2
import sqlalchemy
import pandas as pd
from pandas.io.sql import read_sql, to_sql

# CREATE ENGINE
# engine = sqlalchemy.create_engine("CONNECTION_STRING")

# create datetimes with timezone that have different offsets
data = [[datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60, name=None))],
        [datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=120, name=None))]]

df = pd.DataFrame(data, columns = ['ts_tz'])

# save data - no exceptions occur
to_sql(frame = df, 
       name = 'timezone_diff_offsets_test',
       con = engine,
       schema = 'public',
       if_exists = 'replace',
       index = False)

# read data - offsets are dropped and dtype of column ts_tz is datetime64[ns, UTC]
df = read_sql('SELECT * FROM timezone_diff_offsets_test', engine)
print(df,'\n')
print(df.dtypes)
```

                          ts_tz
    0 2019-11-14 15:12:00+00:00
    1 2019-08-07 13:37:04+00:00 
    
    ts_tz    datetime64[ns, UTC]
    dtype: object

Commit:
handle column with datetimes with tz and diff offsets

detect columns containing only datetimes/NULL recognized as object with function _is_datetime_column_with_dtype_object and coerce them to UTC for read_sql and to_sql

Commit:
avoid collateral damages: only detect diff offsets

Commit:
code formatting + improve comments

jbrockmendel commented:
If the user gives a series of datetimes with mixed timezones, presumably they did it on purpose.  So we should try to preserve that information rather than convert to UTC.  Or am I missing something here?

ThibTrip commented:
> If the user gives a series of datetimes with mixed timezones, presumably they did it on purpose. So we should try to preserve that information rather than convert to UTC. Or am I missing something here?

I don't think that we are really losing information here (see sections to_sql and read_sql below and please correct me if I'm wrong).

# to_sql
Mixed offsets and actually mixed timezones will prevent to_sql from working because the function io.sql._sqlalchemy_type [uses the .dt acessor](https://github.com/pandas-dev/pandas/blob/master/pandas/io/sql.py#L972-L978) and this raises a ValueError with mixed timezones. Since the offsets are not saved in the database anyways (for postgres and SQLite3 AFAIK) converting to UTC should solve the issue without changes for the user.

# read_sql
If all datetimes are in the same timezone and offset then we get a Series with dtype datetime64 and no offset.
If not - for instance in my case one datetime is in winter time and one is in summer time so that makes when psycopg2 adds the offset to my local timezone (Europe/Paris) we get two different offsets (+1 and +2 hours) - then pandas does not convert it to UTC.
I find this rather confusing and I think the example below illustrates this:

<details>
 <summary>Example</summary>

```python
import datetime
import psycopg2
import sqlalchemy
import pandas as pd
import pytz
from pandas.io.sql import read_sql, to_sql

# CREATE ENGINE
# engine = sqlalchemy.create_engine("CONNECTION_STRING")

# create datetimes with timezone that have different offsets
data = [[datetime.datetime(2019, 11, 14, 15, 12, 0, tzinfo = datetime.timezone.utc),
         datetime.datetime(2019, 11, 14, 15, 12, 0, tzinfo = datetime.timezone.utc)],
        [datetime.datetime(2019, 11, 7, 13, 37, 4,  tzinfo = datetime.timezone.utc),
        datetime.datetime(2019, 8, 14, 15, 12, 0, tzinfo = datetime.timezone.utc)]]

df = pd.DataFrame(data, columns = ['ts_tz','ts_tz2'])

# save data - no exceptions occur
to_sql(frame = df, 
       name = 'timezone_diff_offsets_test',
       con = engine,
       schema = 'public',
       if_exists = 'replace',
       index = False)

# read data - offsets are dropped and dtype of column ts_tz is datetime64[ns, UTC]
df = read_sql('SELECT * FROM timezone_diff_offsets_test', engine)
print(df,'\n')
print(df.dtypes)
```
# Output of master branch

                          ts_tz                     ts_tz2
    0 2019-11-14 15:12:00+00:00  2019-11-14 16:12:00+01:00
    1 2019-11-07 13:37:04+00:00  2019-08-14 17:12:00+02:00 

    ts_tz     datetime64[ns, UTC]
    ts_tz2                 object
    dtype: object

# Output of my branch (bugfix_sql_different_offsets)
                          ts_tz                    ts_tz2
    0 2019-11-14 15:12:00+00:00 2019-11-14 15:12:00+00:00
    1 2019-11-07 13:37:04+00:00 2019-08-14 15:12:00+00:00 

    ts_tz     datetime64[ns, UTC]
    ts_tz2    datetime64[ns, UTC]
    dtype: object

</details>

WillAyd commented:
Can you add test cases? It's a little unclear what this is doing without those.

With respect to writing mixed timezones, I'm not sure about this. I actually think the ValueError getting raised makes more sense

mroeschke commented:
@jbrockmendel we have a president of converting timezones (which should include mixed) to UTC when reading from DBs. https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#datetime-data-types

When writing mixed timezones, I agree with @WillAyd in that raising an error is preferable to coercing to UTC if we can't maintain the mixed timezones.

mroeschke commented on pandas/io/sql.py:
```diff
@@ -114,6 +146,10 @@ def _parse_date_columns(data_frame, parse_dates):
+        # handle datetime Series with different offsets
+        # see GH30207
+        if _is_datetime_series_with_different_offsets(df_col):
+            data_frame[col_name] = to_datetime(df_col, utc=True)
```
Just pass `utc=True` into `data_frame[col_name] = _handle_date_column(df_col, format=fmt)` above.

ThibTrip replied:
Yes you are right it should be utc=True. I suppose this was already infered by pandas but we should always enforce it because as you mentionned:
> When reading TIMESTAMP WITH TIME ZONE types, pandas will convert the data to UTC. [source](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#datetime-data-types)

Please note that we still need the logic from L149 to L152 btw it should be "**elif** _is_datetime_series_with_different_offsets" (I shall make a commit soon with this correction and also the utc=True from above) in the case where we have different offsets (with postgres psycopg2 adds the offset to the local timezone which we do not want anyways) because is_datetime64tz_dtype(df_col) is False in this case.

mroeschke commented on pandas/io/sql.py:
```diff
@@ -97,6 +97,38 @@ def _handle_date_column(col, utc=None, format=None):
             return to_datetime(col, errors="coerce", format=format, utc=utc)
 
 
+def _is_datetime_series_with_different_offsets(df_col):
```
I think `infer_dtype` can detect whether the dtype should be upcast to `datetimetz`. Otherwise the logic should live there.

mroeschke replied:
Or use `maybe_convert_objects`

ThibTrip replied:
I have tested infer_dtype, maybe_convert_objects and is_datetime64tz_dtype, and my function _is_datetime_series_with_different_offsets. Only my function works. So we should keep it unless someone has a better idea.

<details>
<summary>Test of infert_dtype, maybe_convert_objects, is_datetime64tz_dtype, _is_datetime_series_with_different_offsets</summary>

```python
from pandas.api.types import infer_dtype
from pandas._libs.lib import maybe_convert_objects
from pandas.api.types import is_datetime64tz_dtype
from pandas.io.sql import _is_datetime_series_with_different_offsets
import pandas as pd
import psycopg2
import datetime
import numpy as np


# this is the kind of data we receive from a psycopg2 cursor when we have a timestamptz column in a postgres table
data_with_offsets = [datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60)),
                     datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=120))]

data_without_offsets = [datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60)),
                        datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60))]


print('infer_dtype with offsets')
print('as list:',infer_dtype(data_with_offsets))
print('as Series:',infer_dtype(pd.Series(data_with_offsets)),'\n')

print('infer_dtype without offsets')
print('as list:',infer_dtype(data_without_offsets))
print('as Series:',infer_dtype(pd.Series(data_without_offsets)),'\n')

print('maybe_convert_objects (only works with numpy array) with offsets')
print('dtype='+str(maybe_convert_objects(np.array(data_with_offsets)).dtype),'\n')

print('maybe_convert_objects without offsets')
print('dtype='+str(maybe_convert_objects(np.array(data_without_offsets)).dtype),'\n')

print('is_datetime64tz_dtype (used in _parse_columns for read_sql) with offsets')
print('as list:',is_datetime64tz_dtype(data_with_offsets))
print('as Series:',is_datetime64tz_dtype(pd.Series(data_with_offsets)),'\n')

print('is_datetime64tz_dtype without offsets')
print('as list:',is_datetime64tz_dtype(data_without_offsets))
print('as Series:',is_datetime64tz_dtype(pd.Series(data_without_offsets)),'\n')

print('_is_datetime_series_with_different_offsets with offsets')
print(_is_datetime_series_with_different_offsets(pd.Series(data_with_offsets)),'\n')

print('_is_datetime_series_with_different_offsets without offsets')
print(_is_datetime_series_with_different_offsets(pd.Series(data_without_offsets)))
```
```
infer_dtype with offsets
as list: datetime
as Series: datetime 

infer_dtype without offsets
as list: datetime
as Series: datetime64 

maybe_convert_objects (only works with numpy array) with offsets
dtype=object 

maybe_convert_objects without offsets
dtype=object 

is_datetime64tz_dtype (used in _parse_columns for read_sql) with offsets
as list: False
as Series: False 

is_datetime64tz_dtype without offsets
as list: False
as Series: True 

_is_datetime_series_with_different_offsets with offsets
True 

_is_datetime_series_with_different_offsets without offsets
False
```

</details>

jreback replied:
pls don't re-invent the wheel, @mroeschke is correct here, if infer_dtype is not doing its job exactly right then then let's fix that, rather than creating an all new *mostly* untested code path.

ThibTrip replied:
If we want to change infer_dtype we'd have to add an additional possible return value such as "datetimetz" because it currently does not differentiate between naive and timezone aware datetimes (see example below). And I don't think we want to convert postgres naive datetimes into UTC.

```python
from pandas.api.types import infer_dtype
import datetime

assert infer_dtype([datetime.datetime(2019,1,1,0,0,0, tzinfo = datetime.timezone.utc)]) == 'datetime'
assert infer_dtype([datetime.datetime(2019,1,1,0,0,0)]) == 'datetime'
```
Should we do that? Otherwise I see 2 other possibilities:

1. change is_datetime64tz_dtype so it returns True even with different offsets
2. simplify the logic in _is_datetime_series_with_different_offsets. This would be sufficient (if we do this we could get rid of the function and do it directly in _parse_date_columns):

```python
if infer_dtype(df_col) == 'datetime' and df_col.dtype == 'object':
    return True
```

jreback replied:
i would rather infer_dryoe return a new type (option 1)

we have extensive test coverage in this routine so adding this would be ok

mroeschke replied:
Yeah I think `infer_dtype` should be able to return `"datetimetz"`, somewhere around here.
https://github.com/pandas-dev/pandas/blob/7ab73a93c94e6557e5d77b2966549955e5c51cbc/pandas/_libs/lib.pyx#L1290

ThibTrip replied:
I have made a new commit with an update on infer_dtype (it now separates between datetime and datetimetz). It seems like quite a big change (infer_dtype is used in lots of places).

I had to modify a method of SQLTable : _sqlalchemy_type to include this "datetimetz". Also in the same function the error message when using to_sql with different offsets/timezones changed so I had to kinda reeinvent the wheel there :crying_cat_face:

Commit:
further modificatons for PR30492 (different offsets sql, see description)

1) improve docstring of _is_datetime_series_with_different_offsets
2) in _parse_date_columns enforce utc if the dtype of df_col is recognized as datetime with tz
3) corect mistake in _parse_date_columns (elif _is_datetime_series_with_different_offsets)
4) make to_sql raise again with diff timezones/offsets by deleting "except ValueError as e" block (L1015-L1025)

ThibTrip commented:
So I have pushed another commit with the suggestions from @mroeschke.

I'd gladly add tests but I am really lost in test_sql.py I don't even know where to start. I suppose I could add tests in one of the functions [here](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_sql.py#L1382-L1487). It would be easier if I had access to the database (of if I could use my own for local testing). Maybe I just need some more time to figure this out but if some wants to help I thought of some testing (it passed) with postgres already (see below).

<details>
<summary>Additional tests</summary>

```python
import datetime
import psycopg2 
import sqlalchemy
import pandas as pd
import pytz
from pandas.io.sql import read_sql, to_sql

# CREATE ENGINE
# engine = sqlalchemy.create_engine("CONNECTION_STRING")

# create test data
df = pd.DataFrame()

## CASE 1 (all UTC)
## no exception should occur with to_sql, 
## it should be saved as timestamp with timezone (where compatible e.g. postgres)
## read_sql should give back datetime64[ns, UTC] 
## and not object dtype (discard local offsets winter/summer time)
df['ts_utc_winter_summer'] = [datetime.datetime(2019, 11, 7, 13, 37, 4,  
                                                tzinfo=datetime.timezone.utc),
                              datetime.datetime(2019, 8, 14, 15, 12, 0, 
                                                tzinfo=datetime.timezone.utc)]

## CASE 2 (different offsets)
## to_sql should raise (it did with master already)
df['ts_diff_offset'] = [datetime.datetime(2019, 11, 14, 16, 12, 0, 
                             tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60)),
                        datetime.datetime(2019, 8, 7, 15, 37, 4, 
                             tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=120))]

## CASE 3 (different timezones)
## to_sql should raise (it did with master already)
df['ts_diff_tz'] = [datetime.datetime(2019, 11, 14, 16, 12, 0, 
                                   tzinfo=datetime.timezone.utc),
                    datetime.datetime(2019, 11, 14, 16, 12, 0, 
                                   tzinfo=pytz.timezone('Europe/Paris'))]

## CASE 4 (naive)
## no exception should occur with to_sql, 
## it should be saved as timestamp without timezone
## read_sql should give back datetime64[ns] (it did with master already)
df['ts_naive'] = [datetime.datetime(2019, 11, 7, 13, 37, 4),
                  datetime.datetime(2019, 8, 14, 15, 12, 0)]

# show df
pd.options.display.width = 500
print(df,'\n')

# make sure we have the expected types
# where different timezones/offsets we should have an object dtype
expected_dtypes = {'ts_utc_winter_summer': 'datetime64[ns, UTC]',
                   'ts_diff_offset': 'O',
                   'ts_diff_tz': 'O',
                   'ts_naive': '<M8[ns]'}

for col in df.columns:
    assert df[col].dtype == expected_dtypes[col]

# try to_sql one column at a time (as a DataFrame subset not a Series)
for col in df.columns:
    try:
        to_sql(frame = df[[col]], 
               name = 'timezone_tests',
               con = engine,
               schema = 'public',
               if_exists = 'replace',
               index = False)

        df_db = read_sql('SELECT * FROM timezone_tests', engine)

        if col == 'ts_naive':
            assert df_db['ts_naive'].dtype == 'datetime64[ns]'
        else:
            assert df_db[col].dtype == 'datetime64[ns, UTC]'
    except ValueError as e:
        print(f"Error with column {col}:", e)
```
```
       ts_utc_winter_summer             ts_diff_offset                 ts_diff_tz            ts_naive
0 2019-11-07 13:37:04+00:00  2019-11-14 16:12:00+01:00  2019-11-14 16:12:00+00:00 2019-11-07 13:37:04
1 2019-08-14 15:12:00+00:00  2019-08-07 15:37:04+02:00  2019-11-14 16:12:00+00:09 2019-08-14 15:12:00 

Error with column ts_diff_offset: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True
Error with column ts_diff_tz: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True
```

</details>

WillAyd commented:
I think can just add the new tests to `_TestSQLApi`

Commit:
do not enforce utc for columns in parse_dates

ThibTrip commented:
It seems like local testing with SQL is not possible (it always fails because it won't connect).

But I think I have found the culprit. By adding utc=True in function _parse_date_columns (see comment of @mroeschke [here](https://github.com/pandas-dev/pandas/pull/30492/files/66ebe355c9c080536574ad8c5f0a4a3cb5e5be6b)) we changed the behavior with the argument "parse_dates". See example below:

<details>
<summary>Example</summary>

```python
from sqlalchemy import create_engine
import pandas as pd
import datetime
engine = create_engine('sqlite:///sqlite_test.sqlite3')

# create and save test data
data = [datetime.datetime(2019, 11, 11, 9, 0, 0,  
                          tzinfo=datetime.timezone.utc),
        datetime.datetime(2019, 6, 1, 9, 0, 0, 
                          tzinfo=datetime.timezone.utc)]

df = pd.DataFrame(data, columns = ['ts'])
df.to_sql('test_utc_true', con = engine, index = False, if_exists = 'replace')

# with utc=True change
df = pd.read_sql('SELECT * FROM test_utc_true', 
                 con = engine, 
                 parse_dates = ['ts'])
print(df,'\n')
print(df.dtypes)
```
```
                         ts
0 2019-11-11 09:00:00+00:00
1 2019-06-01 09:00:00+00:00 

ts    datetime64[ns, UTC]
dtype: object
```
```python
# without utc=True change
df = pd.read_sql('SELECT * FROM test_utc_true', 
                 con = engine, 
                 parse_dates = ['ts'])
print(df,'\n')
print(df.dtypes)
```
```
                   ts
0 2019-11-11 09:00:00
1 2019-06-01 09:00:00 

ts    datetime64[ns]
dtype: object
```

</details>

With the consideration from above I made a separate block for the columns in parse_dates in a new comit.

Commit:
separate between "datetime" and "datetimetz" in infer_dtype

Commit:
add tests for updated infer_dtype

Commit:
use updated infer_dtype

Commit:
better error message if attempting to save different timezones/offsets

Commit:
fix trailing whitespace

Commit:
fix import issue -> use pytz

Commit:
reflect update on infer_dtype

mroeschke commented on pandas/_libs/lib.pyx:
```diff
@@ -1213,6 +1213,11 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     'datetime'
 
+    >>> import pytz
+    >>> infer_dtype([datetime.datetime(2013,1,1,0,0,0,
```
Let's just make the example with `pd.Timestamp(..., tz='UTC')`

mroeschke commented on pandas/_libs/lib.pyx:
```diff
@@ -1732,6 +1740,31 @@ def is_datetime_with_singletz_array(values: ndarray) -> bool:
     return True
 
 
+def is_datetime_with_tz_array(values: ndarray) -> bool:
```
Could you create a `DatetimeTZValidator` in the same vein as `DatetimeValidator` in the same file instead?

ThibTrip replied:
I will look into that, also in the example below the infered dtype is "datetime64" although it should be "datetimetz" I think.

```python
s = pd.Series(
    [
        pd.Timestamp("2013-01-01 13:00:00-0800", tz="US/Pacific"),
        pd.Timestamp("2013-01-02 14:00:00-0800", tz="US/Pacific"),
    ]
)
assert s.dtype == "datetime64[ns, US/Pacific]"
assert infer_dtype(s, skipna=True) == "datetime64"
```

jreback replied:
note that we actually might want 2 different returns here (this is a general statement)

1) datetimetz_same - all the entries are the same tz
2) datetimetz_mixed - all entries have a tz but can be different - note that mixing a date time with no time zone would not be valid here (and ultimately would be mixed)

ThibTrip replied:
@mroeschke I made a DatetimeTZValidator (according to my tests -see code block after the one below- it works).

```
cdef class DatetimeTZValidator(TemporalValidator):
    cdef bint is_value_typed(self, object value) except -1:
        return PyDateTime_Check(value) and (value.tzinfo is not None)

    cdef inline bint is_valid_null(self, object value) except -1:
        return is_null_datetime64(value)
```

There is still one point to clear concerning infer_dtype: as much as I like the idea from @jreback I think the distinction between datetime64 and datetime is more important right now and that we need to add a "datetime64tz" return value for inder_dtype. Later on we could then implement "datetimetz_same", "datetimetz_mixed", "datetimet64tz_same", "datetime64tz_mixed".

I prepared some tests and checked if I was able to modify pandas._libs.lib to achieve the distinctions "datetimetz"/"datetime64tz"/"datetime"/"datetime64" and actually got it to work. There is definitely room for improvements and I will make a commit later (unless you do not like this idea). Anyhow here are the tests I checked:

```python
from pandas.api.types import infer_dtype
import pandas as pd
import numpy as np
import datetime

assert infer_dtype([datetime.datetime(2019,1,1,0,0,0, 
                                      tzinfo = datetime.timezone.utc)]) == 'datetimetz'
assert infer_dtype([pd.Timestamp('2019-01-01', tz = 'UTC')]) == 'datetimetz'
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc)]) == 'datetimetz'

# mixed (with and without tz will return datetime)
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc),
                    pd.Timestamp('2019-01-01')]) == 'datetime'
assert infer_dtype([datetime.datetime(2019,1,1,0,0,0)]) == 'datetime'
assert infer_dtype([pd.Timestamp('2019-01-01')]) == 'datetime'

# this monstrocity will return "datetimetz"
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc),
                    datetime.datetime(2019,1,1,0,0,0, 
                                      tzinfo = datetime.timezone.utc)]) == "datetimetz"

s = pd.Series([pd.Timestamp("2013-01-01 13:00:00-0800", 
                            tz="US/Pacific"),
               pd.Timestamp("2013-01-02 14:00:00-0800", 
                            tz="US/Pacific")])
assert infer_dtype(s, skipna=True) == "datetime64tz"

s = pd.Series([pd.Timestamp("2013-01-01"),
               pd.Timestamp("2013-01-02")])
assert infer_dtype(s, skipna=True) == "datetime64"

arr = np.array([datetime.datetime(2019,1,1,0,0,0, 
                                  tzinfo = datetime.timezone.utc),
                datetime.datetime(2019,1,1,0,0,0, 
                                  tzinfo = datetime.timezone.utc)])
assert infer_dtype(arr, skipna=True) == "datetime64tz"
```

Also I wanted to say sorry for the mess I still need to learn this is my first PR and first time really working with Cython. I am focusing on infer_dtype before fixing the errors in the Azure pipelines (I already observed them though).

jreback replied:
don’t distinguish datetime64 ; these are just things that are convertible to date time dtypes

make your changes S simple as possible

mroeschke commented on pandas/io/sql.py:
```diff
@@ -108,13 +108,18 @@ def _parse_date_columns(data_frame, parse_dates):
+        # this will handle datetime Series with different offsets
+        # see GH30207
+        if lib.infer_dtype(df_col) == "datetimetz":
+            data_frame[col_name] = to_datetime(df_col, utc=True)
```
As mentioned before just use `_handle_date_column(df_col, utc=lib.infer_dtype(df_col) == "datetimetz", format=fmt)` or something similar.

ThibTrip replied:
Oh ok I did not get this before. So since [is_datetime64tz_dtype](https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/core/dtypes/common.py#L539-L574) is used to recognize the extension dtype [DatetimeTZDtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeTZDtype.html) specifically and since a Series with mixed timezones/offsets will result in a dtype "object" I would propose to modify the function as such (to me this seems close to what you wrote):

```
def _parse_date_columns(data_frame, parse_dates):
    """
    Force non-datetime columns to be read as such.
    Supports both string formatted and integer timestamp columns.
    """
    parse_dates = _process_parse_dates_argument(parse_dates)

    # we want to coerce datetime64_tz dtypes for now to UTC
    # we could in theory do a 'nice' conversion from a FixedOffset tz
    # GH11216
    for col_name, df_col in data_frame.items():
        # infer_dtype == "datetimetz" will work with different timezones/offsets
        # (in such cases the dtype is object and not DatetimeTZDtype)
        is_datetimetz = (
            is_datetime64tz_dtype(df_col) or lib.infer_dtype(df_col) == "datetimetz"
        )
        if is_datetimetz or col_name in parse_dates:
            try:
                fmt = parse_dates[col_name]
            except TypeError:
                fmt = None
            data_frame[col_name] = _handle_date_column(
                df_col, utc=is_datetimetz, format=fmt
            )
    return data_frame
```

ThibTrip replied:
As @jreback mentionned let's do this in another PR :+1:

mroeschke commented on pandas/io/sql.py:
```diff
@@ -966,16 +971,35 @@ def _sqlalchemy_type(self, col):
+                # raise a better error message than
+                # "'Series' object has no attribute 'tz'"
+                except AttributeError as e:
+                    get_offset = lambda dt: (
```
I think we can only hit this path if we have multiple timezones right? In that case I'd just re-raise a `ValueError` here and skip all this inference.

jreback reviewed (changes requested):
@ThibTrip i would prefer that you just do the *infer_dtype* change in this PR, then submit a followup after that is merged for the sql changes. This is going to be tricky as-is.

Commit:
use a DatetimeTZValidator, update infer_dtype to make it pass all tests

Commit:
update test after infer_dtype modification

Commit:
revert sql to its master state

Commit:
change sql.py only for updated infer_dtype

Commit:
fix linting issue

Commit:
make another update due to infer_dtype

Commit:
fix linting issue (black test_inference.py)

ThibTrip commented:
@jreback should I rename the title of the PR to something like:

"BUG: sql different offsets (1/2 modify infer_dtype)"

then I can call the next PR:

"BUG: sql different offsets (2/2 modify sql.py)"

Also I think whatsnew entry does not apply for this PR right? infer_dtype is not part of the public API. Is there something else to do from my part for this PR?

jreback commented:
> @jreback should I rename the title of the PR to something like:
> 
> "BUG: sql different offsets (1/2 modify infer_dtype)"
> 
> then I can call the next PR:
> 
> "BUG: sql different offsets (2/2 modify sql.py)"
> 
> Also I think whatsnew entry does not apply for this PR right? infer_dtype is not part of the public API. Is there something else to do from my part for this PR?

yes pls rename this one and *only* include infer_dtype changes

ThibTrip commented:
Hey, I was wondering why my PR has not been merged into master yet. I don't mean to rush anything but if there is anything else I have to do please tell me :+1:.
I suppose everyone is very busy with version 1 which is very cool :rocket:.

jreback commented:
merge master

we have so many PRs if u want a look you can ping (like u did)

Commit:
Merge branch 'master' into bugfix_sql_different_offsets

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1240,6 +1243,10 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     elif hasattr(value, 'dtype'):
         # this will handle ndarray-like
         # e.g. categoricals
+        # begin by checking if there is a tz attribute
```
hmm, this is not the right place for this

ThibTrip replied:
Ok so I thought about it again and prepared some tests.

<details>
 <summary>Tests</summary>

```python
import numpy as np
import pandas as pd
import pytz
import datetime
from pandas._libs.lib import infer_dtype

tests = [pd.Series([pd.Timestamp('2013-01-01', tz='UTC'), np.datetime64('NaT')]),
         pd.Series([pd.Timestamp('2013-01-01', tz='UTC')]),
         pd.Series([pd.Timestamp('2013-01-01', tz='America/Toronto'), pd.Timestamp('2013-01-01', tz='Europe/Paris')]),
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), np.datetime64('NaT')],
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), None],
         [datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc), datetime.datetime(2018, 1, 1, tzinfo=pytz.timezone('America/Toronto'))],
         np.array(datetime.datetime(2018, 1, 1, tzinfo=datetime.timezone.utc))]

for i, test in enumerate(tests):
    print(f'running test {i}')
    assert infer_dtype(test) == 'datetimetz'

```
</details>

And you are right I think the logical place is actually after the "except TypeError" for extension arrays and right before "_try_infer_map". By the way if you removed this modification of mine (where I check the "tz" attribute) then only the second test would actually fail.

Anyhow here is the code in the new place and all tests pass:

```python
        except TypeError:
            # This gets hit if we have an EA, since cython expects `values`
            #  to be an ndarray
            # before using try_infer_map check if there is a tz attribute
            # because try_infer_map cannot be used to differentiate dtypes
            # such as "datetime64[ns]" and "datetime64[ns, {timezone}]"
            if hasattr(value.dtype, 'tz'):
                return "datetimetz"
            value = _try_infer_map(value)
```
Btw if you are wondering why I have not modified "_try_infer_map" here is an explanation below.

<details>

<summary>why I have not changed _try_infer_map</summary>

So in the case of the second test the dtype is "datetime64[ns, UTC]" and the attribute "kind" of the dtype is "M" which is in the mapper "_TYPE_MAP"  so the value gets picked up by the function "_try_infer_map" returning "datetime64". 

Adding "datetime64[ns, UTC]"  to the mapper does not seem like a good solution to me since we would need to add every other timezones e.g. "datetime64[ns, Europe/Paris]" (I tried it and the same scenario happens).

Of course we could make some kind of rule in _try_infer_map I suppose (name of dtype starts with datetime64 etc.) but that does not seem very elegant to me either and does not seem to fit with the purpose of the function.

</details>

jreback replied:
you need to move this

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1318,7 +1329,11 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
         if is_datetime_array(values):
-            return "datetime"
+            # only check for non nulls to avoid numpy nat
+            if is_datetimetz_array(values[~isnaobj(values)]):
```
same

ThibTrip replied:
Sorry not sure what the problem is here. I will investigate later as it is getting late :sleeping:. If you can tell me a bit more why you think that it would most likely help my research. Thanks!

jreback reviewed (changes requested):
pls rebase as well. pls grep the codebase for infer_dtype when the result is a 'datetime64' I can't believe you don't have to change some of these.

Commit:
Merge branch 'master' into bugfix_sql_different_offsets

ThibTrip commented:
I merged master again (commit id ca3bfcc54676b04ac5150c913328d26920bb2591) and cythonized again so everything should be up to date. I don't understand why I had an issue with two of the CI tests. Seems to be related to tab completion. I will investigate but I don't really see any relation with what I did... Also @jreback I am leaving you some notes on your changes proposals.

jbrockmendel commented:
Tab completion is unrelated, don’t worry about it

On Sun, Jan 26, 2020 at 12:09 PM ThibTrip <notifications@github.com> wrote:

> I merged master again (commit id ca3bfcc
> <https://github.com/pandas-dev/pandas/commit/ca3bfcc54676b04ac5150c913328d26920bb2591>)
> and cythonized again so everything should be up to date. I don't understand
> why I had an issue with two of the CI tests. Seems to be related to tab
> completion. I will investigate but I don't really see any relation with
> what I did... Also @jreback <https://github.com/jreback> I am leaving you
> some notes on your changes proposals.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/30492?email_source=notifications&email_token=AB5UM6HKGK646SN74HKZAMTQ7XUXPA5CNFSM4J7PKTG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ54UIA#issuecomment-578538016>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5UM6DIQSRRVMB2TZPLU5LQ7XUXPANCNFSM4J7PKTGQ>
> .
>

Commit:
Merge branch 'master' into bugfix_sql_different_offsets

Commit:
improve placement of tz attr check, use function for datetimetz with nan

ThibTrip commented:
So I realized one check did not seem necessary as I was able to pass all my tests (plus my other tests I already showed you with psycopg2) and pandas tests (I mean apart from the tab completion and also I made a bad indentation somewhere it seems) without it. So I just removed it. I think I added it to fix something then realized the trigger was elsewhere and then forgot about it.

I also created and used a function to check for datetimetz with nan (got inspired by the integer na funtion). I hope this will seem better to you because I am running out of ideas 😞

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1284,11 +1287,18 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
         except TypeError:
             # This gets hit if we have an EA, since cython expects `values`
             #  to be an ndarray
+            # before using try_infer_map check if there is a tz attribute
```
this doesn't belong here at all, this is only for handling a specific dtype w/o special cases, which is what you are doing here.

ThibTrip replied:
Yes I think I get the idea. I moved the timezone aware Series check above in my new commit.

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1362,7 +1372,11 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     elif PyDateTime_Check(val):
         if is_datetime_array(values):
-            return "datetime"
+            if (is_datetimetz_na_array(values)
```
you just need is_datetimetz_array

ThibTrip replied:
Sorry this was quite stupid the part <code>or is_datetime_tz_array(values)</code> would never be executed. I have fixed this.

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1740,6 +1754,36 @@ cpdef bint is_datetime_array(ndarray values):
+    return validator.validate(values)
+
+
+cdef class DatetimeTZNaValidator(TemporalValidator):
```
you don't need this at all

ThibTrip replied:
Yes I realize no this is redundant the DatetimeTZValidator works fine with missing values (and we don't need to differentiate between datetimetz and datetimetz with na either).

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1740,6 +1754,36 @@ cpdef bint is_datetime_array(ndarray values):
+        return is_null_datetime64(value)
+
+
+cpdef bint is_datetimetz_na_array(ndarray values):
```
you don't need this

ThibTrip replied:
I removed this in my new commit.

jreback commented on pandas/io/sql.py:
```diff
@@ -963,7 +963,7 @@ def _sqlalchemy_type(self, col):
         )
 
-        if col_type == "datetime64" or col_type == "datetime":
+        if col_type in ("datetime64", "datetime", "datetimetz"):
```
can you revert all of these changes in this file

ThibTrip replied:
I think we need this to pass the tests. As you can see there were some failures. I will try some tests locally (I have to copy paste them since I am not sure who to change the configuration for the connection strings) and give you a new feedback when I'm done :).

jreback commented on pandas/tests/dtypes/test_inference.py:
```diff
@@ -794,6 +794,12 @@ def test_infer_dtype_datetime(self):
             arr = np.array([n, datetime(2011, 1, 1)])
             assert lib.infer_dtype(arr, skipna=True) == "datetime"
 
+            arr = np.array([n, datetime(2011, 1, 1, tzinfo=pytz.utc)])
```
can you move these to a new test

ThibTrip replied:
Done. I took the liberty of adding some more tests for cases that seem relevant to me. I hope that's alright.

Commit:
Merge branch 'master' into bugfix_sql_different_offsets

Commit:
move tz aware Series check to its own code block

Commit:
Remove DatetimeTZNaValidator and is_datetimetz_na_array

Commit:
Revert changes to sql.py

Commit:
separate tests for datetimetz inference + add tests

Commit:
fix linting

ThibTrip commented:
Hey, I wanted to say thanks for your patience jreback. 

This PR has been and is quite difficult for me and if anyone wants to help in my branch or even take over I do not mind at all. Not that I want to give up or anything this is actually very interesting!

Anyhow after testing minutely the sql.py module for the failures of the last CI run I can confirm that we need to check if the infered col_type is equal to "datetimetz" in method **SQLTable._sqlalchemy_type** otherwise **test_sqlalchemy_type_mapping** and **test_datetime_with_timezone_roundtrip** (both in pandas/tests/io/test_sql.py) will fail with an sqlalchemy engine connected to a SQLite or postgres database (because the sql dtype of the table model will be Text instead of TIMESTAMP).

However the other bit of code I changed in SQLiteTable._sql_type_name where I also check if the infered col_type is equal to "datetimetz" does not seem to be necessary to pass tests. But it would not give the expected data type in the table model so it seems logical to me to do this change as well (see SQLite test details).

<details>

<summary>SQLite test</summary>

```python
import sqlite3
from pandas import DataFrame, to_datetime, date_range
from pandas.io import sql
from sqlalchemy import create_engine

# with engine = sqlite3.connect(":memory:") we get the same output
engine = create_engine('sqlite:///:memory:') 
db = sql.SQLDatabase(engine)
expected = DataFrame({"A": date_range("2013-01-01 09:00:00", 
                                      periods=3, tz="US/Pacific")})
table = sql.SQLiteTable("test_type", db, frame=expected)
# if we do not check for "datetimetz" in SQLiteTable._sql_type_name
# we end up with "A" as "TEXT" instead of "TIMESTAMP" (see table below)
table.table 
```

```python
['CREATE TABLE "test_type" (\n"index" INTEGER,\n  "A" TEXT \n)',
 'CREATE INDEX "ix_test_type_index"ON "test_type" ("index")']
```


</details>

How shall I proceed? Thanks in advance.

mroeschke commented:
Hi @ThibTrip.

Sorry your PR had went unnoticed for a good period of time. Do you have time to merge master?

I suspect we need to add that `col_name == 'datetimetz'` check somewhere.

Commit:
Merge branch 'master' into bugfix_sql_different_offsets

ThibTrip commented:
Hi @mroeschke! Not a problem at all! Also sorry I took so long to reply I have been very busy with work and personal projects. I have merged master. I had a check for datetimetz but I removed it (we need it at the very least here: https://github.com/pandas-dev/pandas/blob/master/pandas/io/sql.py#L973) because the idea IIRC was to touch as little files as possible and do the change in two PRs (in this one we would only make sure infer dtypes recognizes we are in presence of datetime with timezone).
I am going to add it back in another commit, test what I can locally (sadly I never figured out how to test SQL locally) and then we'll see.

Commit:
add column type checks to pass tests

ThibTrip commented:
@mroeschke I managed to test SQL locally :+1: ! It was not complicated at all I just did not have sufficient knowledge back when I started this PR. I just needed to install postgres and create a database pandas_nosetest (in the tests it connects with the default postgres user on localhost).

Anyhow I have lots of errors locally that I did not have before and in other tests modules so I am rather surprised. Some tests like pandas.tests.plotting.test_datetimelike.TestTSPlot seem completely irrelevant to my PR yet still fail. So I pushed to see if I have the same errors in the CI tools.

Commit:
move detection of DatetimeTZDtype

ThibTrip commented:
@jreback I think you mentioned the section about DatetimeTZDtype detection before and I did not know how to place it otherwise back then. Sorry about that. I think I know now what you had in mind and have pushed a new commit. Please tell me if that's god like this :+1: .

mroeschke commented on pandas/tests/dtypes/test_inference.py:
```diff
@@ -834,6 +834,30 @@ def test_infer_dtype_datetime(self):
 
+    def test_infer_dtypes_datetimetz(self):
+
+        # starts with nan
```
Can you split this into 3 different tests?

And for these `# starts with nan` test, could you use `pytest.mark.parameterize`

mroeschke commented:
Just one comment on the tests. The changes to `sql.py` look good.

I'll defer to  @jreback on `infer_dtype`.

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1283,6 +1286,11 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     if util.is_array(value):
         values = value
+
+    elif hasattr(value, 'dtype') and hasattr(value.dtype, 'tz'):
```
doesn't belong here at all

jreback commented on pandas/_libs/lib.pyx:
```diff
@@ -1364,7 +1372,10 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
     elif PyDateTime_Check(val):
         if is_datetime_array(values):
-            return "datetime"
+            if is_datetimetz_array(values):
```
this is likely enough

ThibTrip replied:
I will very simply delete the part above and test locally. I'll check again which paths the tests function take. Perhaps it is not useful or perhaps there is a better way.

ThibTrip replied:
So @jreback I executed my plan and one of the tests failed (see code of failed test below) after I removed the part which is not in the correct place (<code>elif hasattr(value, 'dtype') and hasattr(value.dtype, 'tz'):...</code>).

I followed the path taken during the test. As you can see by the code and my comments below (look for comments with lots of hyphens) it goes through _try_infer_map (see code below) where it is not recognized as "datetimetz" but as "datetime64" (s.dtype.kind is "M" which in _TYPE_MAP translates to "datetime64").

So the kind and base attribute of s.dtype don't help identify our Series as "datetimetz" with try_infer_map (see code below) and since it's a mapping and not say regex the name attribute of s.dtype which is "datetime64[ns, US/Pacific]" won't match (we'd have to include all timezones in the mapper which does not seem like the right solution).

Between this and the fact that afterwards the dtype is supposed to be returned this brings me to the conclusion that the check for DatetimeTZdtype must be right before try_infer_map as it was during a [previous commit](https://github.com/pandas-dev/pandas/pull/30492/commits/da52caf536e3012d786332b79011bee0eed82151#diff-a23dec10f398de8ed7358a4b4eca76c9R1300-R1301) (or we would have to do some surgery and check no other inferences break).


**_failed_test_**
```python
import pandas as pd
from pandas import _lib as lib
from pandas import Series

# pandas/tests/series/test_constructors.py
# TestSeriesConstructors.test_constructor_with_datetime_tz
# inference
s = Series([pd.Timestamp("2013-01-01 13:00:00-0800", tz="US/Pacific"),
            pd.Timestamp("2013-01-02 14:00:00-0800", tz="US/Pacific")])
# Assertion Error 'datetime64' == 'datetimetz'
assert lib.infer_dtype(s, skipna=True) == "datetimetz"
```


**_infer_dtype_**
```python
if util.is_array(value):
    values = value

elif hasattr(value, 'dtype'):  # <------------ GOES HERE
    # this will handle ndarray-like
    # e.g. categoricals
    try:
        values = getattr(value, '_values', getattr(value, 'values', value))
    except TypeError:  # <------------ GOES HERE
        # This gets hit if we have an EA, since cython expects `values`
        #  to be an ndarray
        value = _try_infer_map(value) # -----SEE try_infer_map code below----
        if value is not None:
            return value

        # its ndarray like but we can't handle
        raise ValueError(f"cannot infer type for {type(value)}")

else:
    if not isinstance(value, list):
        value = list(value)
    from pandas.core.dtypes.cast import (
        construct_1d_object_array_from_listlike)
    values = construct_1d_object_array_from_listlike(value)
```
**_try_infer_map_**

```cython
_TYPE_MAP = {'categorical': 'categorical', 'category': 'categorical',
             'int8': 'integer', 'int16': 'integer', 'int32': 'integer',
             'int64': 'integer', 'i': 'integer', 'uint8': 'integer',
             'uint16': 'integer', 'uint32': 'integer', 'uint64': 'integer',
             'u': 'integer', 'float32': 'floating', 'float64': 'floating',
             'f': 'floating', 'complex64': 'complex', 'complex128': 'complex',
             'c': 'complex', 'string': 'string', 'S': 'bytes', 'U': 'string', 
             'bool': 'boolean', 'b': 'boolean', 'datetime64[ns]': 'datetime64',
             'M': 'datetime64', 'timedelta64[ns]': 'timedelta64',
             'm': 'timedelta64', 'interval': 'interval'}

# NOT COPIED: types which only exist on certain platform
# (it's for float16, flat128 and complex256)

cdef _try_infer_map(v):
    """
    If its in our map, just return the dtype.
    """
    cdef:
        object attr, val
    for attr in ['name', 'kind', 'base']:
        # -----------------------------------
        # for our example:
        # name datetime64[ns, US/Pacific]
        # kind M
        # base dtype('<M8[ns]')
        # note: base won't be called bc kind "M" in is _TYPE_MAP
        # -----------------------------------
        val = getattr(v.dtype, attr)
        if val in _TYPE_MAP:
            return _TYPE_MAP[val]
    return None
```

jbrockmendel replied:
can we do this without doing multiple passes through the data?

ThibTrip replied:
@jbrockmendel cython.datetime does not seem to provide a way to distinguish between datetime with and without timezone however this is something we can do I think (I tested it locally and it works): in DatetimeValidator we can check the absence of tzinfo.

```cython
cdef class DatetimeValidator(TemporalValidator):
    cdef bint is_value_typed(self, object value) except -1:
        return PyDateTime_Check(value) and (value.tzinfo is None) # <-----
```

And then we change the part with the repetition to:

```python
    elif PyDateTime_Check(val):
        if is_datetime_array(values):
            return "datetime"
        elif is_datetimetz_array(values):
            return "datetimetz"    
```

Commit:
use pytest parametrize for datetimetz

mroeschke commented on pandas/tests/dtypes/test_inference.py:
```diff
@@ -834,6 +834,31 @@ def test_infer_dtype_datetime(self):
+
+        # pd.Timestamps with different timezones
+        arr = [
+            pd.Timestamp("2011-01-02", tz="UTC"),
```
you can use `pytest.mark.parameterize` over `[datetime, Timestamp]` and construct these like the `datetime` form you have above

ThibTrip replied:
Yep... makes sense! Don't know why I did not think about it! I will make a new commit after I do some tests related to the request of @jreback.

jbrockmendel commented on pandas/_libs/lib.pyx:
```diff
@@ -1742,6 +1753,21 @@ cpdef bint is_datetime_array(ndarray values):
+        return PyDateTime_Check(value) and (value.tzinfo is not None)
+
+    cdef inline bint is_valid_null(self, object value) except -1:
+        return is_null_datetime64(value)
```
we might want to only accept `value is NaT` here

ThibTrip replied:
@jbrockmendel sorry but I think that we should at least test for None as well since infer_dtype will be used on arrays we receive from the various SQL drivers and None is AFAIK the default null value there. As you can see in the example below (this is the issue that we are currently trying to fix) the array has not been parsed to a DatetimeTZDtype and still contains None:

```python
import pandas as pd
from sqlalchemy import create_engine

# REPLACE with valid postgresql connection string
engine = create_engine('postgresql://............')

df = pd.DataFrame({'dt': [pd.Timestamp('2019-11-14 15:12:00+0000', tz='UTC'),
                          pd.Timestamp('2019-08-07 13:37:04+0000', tz='UTC'),
                          pd.NaT]})
df.to_sql('null_value_test', con=engine, schema='tests', if_exists='replace',
          index=False)
pd.read_sql('SELECT * FROM tests.null_value_test', con=engine)
```
|    | dt                        |
|---:|:--------------------------|
|  0 | 2019-11-14 16:12:00+01:00 |
|  1 | 2019-08-07 15:37:04+02:00 |
|  2 |                      None |

jbrockmendel replied:
Thank you for clarifying.  I'll need to give some thought to whether this is the appropriate place to do this check.

jreback commented:
I think these changes are nice, but closing as stale. ping / open a new PR if you want to continue
